libethereum/EthereumPeer.cpp

```diff
@@ -67,17 +54,20 @@ void EthereumPeer::setStatus(unsigned _protocolVersion, u256 const& _networkId,
 std::string EthereumPeer::validate(
     h256 const& _hostGenesisHash, unsigned _hostProtocolVersion, u256 const& _hostNetworkId) const
 {
-    std::string error;
-    if (m_genesisHash != _hostGenesisHash)
-        error = "Invalid genesis hash.";
+    std::stringstream error;
+    if (m_networkId != _hostNetworkId)
+        error << "Network identifier mismatch. Host network id: " << _hostNetworkId
+              << ", peer network id: " << m_networkId;
     else if (m_protocolVersion != _hostProtocolVersion)
-        error = "Invalid protocol version.";
-    else if (m_networkId != _hostNetworkId)
-        error = "Invalid network identifier.";
+        error << "Protocol version mismatch. Host protocol version: " << _hostProtocolVersion
+              << ", peer protocol version: " << m_protocolVersion;
+    else if (m_genesisHash != _hostGenesisHash)
+        error << "Genesis hash mismatch. Host genesis hash: " << _hostGenesisHash
```

Comment:
Too bad this doesn't use `operator<<` overloads for logging, so it outputs long full hash. Maybe change it to outputting `.abridged()` ?